Cross-compile for Spark 2.4.7, 3.0.1

arielshaqed opened the pull request:
Testing:

```
ariels@redqueen:~/dev/spark-client$ SPARK_HOME=$HOME/dev/spark-2.4.7-bin-hadoop2.7   spark-shell --master local --packages io.treeverse:core-247_2.11:0.1.0-SNAPSHOT  --conf spark.hadoop.lakefs.api.url=http://yoni-test.lakefs.dev/api/v1  --conf spark.hadoop.lakefs.api.access_key=AKIA... --conf spark.hadoop.lakefs.api.secret_key=... --conf spark.hadoop.fs.s3a.access.key=AKIA... --conf spark.hadoop-fs.s3a.secret.key=...
Ivy Default Cache set to: /home/ariels/.ivy2/cache
The jars for the packages stored in: /home/ariels/.ivy2/jars
:: loading settings :: url = jar:file:/home/ariels/dev/spark-2.4.7-bin-hadoop2.7/jars/ivy-2.4.0.jar!/org/apache/ivy/core/settings/ivysettings.xml
io.treeverse#core-247_2.11 added as a dependency
:: resolving dependencies :: org.apache.spark#spark-submit-parent-9701fd43-edf3-4326-882f-e66c357e86f5;1.0
        confs: [default]
        found io.treeverse#core-247_2.11;0.1.0-SNAPSHOT in local-ivy-cache
        [0.1.0-SNAPSHOT] io.treeverse#core-247_2.11;0.1.0-SNAPSHOT
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.apache.ivy.util.url.IvyAuthenticator (file:/home/ariels/dev/spark-2.4.7-bin-hadoop2.7/jars/ivy-2.4.0.jar) to field java.net.Authenticator.theAuthenticator
WARNING: Please consider reporting this to the maintainers of org.apache.ivy.util.url.IvyAuthenticator
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
        found com.thesamet.scalapb#scalapb-runtime_2.11;0.9.7 in central
        found com.thesamet.scalapb#lenses_2.11;0.9.7 in central
        found com.google.protobuf#protobuf-java;3.8.0 in central
        found com.lihaoyi#fastparse_2.11;2.1.2 in central
        found com.lihaoyi#sourcecode_2.11;0.1.6 in central
        found org.rocksdb#rocksdbjni;6.6.4 in central
        found commons-codec#commons-codec;1.15 in local-m2-cache
        found org.apache.hadoop#hadoop-aws;2.7.7 in central
        found org.apache.hadoop#hadoop-common;2.7.7 in central
        found org.apache.hadoop#hadoop-annotations;2.7.7 in central
        found commons-cli#commons-cli;1.2 in local-m2-cache
        found org.apache.commons#commons-math3;3.1.1 in central
        found xmlenc#xmlenc;0.52 in local-m2-cache
        found commons-httpclient#commons-httpclient;3.1 in local-m2-cache
        found commons-logging#commons-logging;1.1.3 in local-m2-cache
        found commons-io#commons-io;2.4 in local-m2-cache
        found commons-net#commons-net;3.1 in local-m2-cache
        found commons-collections#commons-collections;3.2.2 in local-m2-cache
        found javax.servlet#servlet-api;2.5 in central
        found org.mortbay.jetty#jetty;6.1.26 in central
        found org.mortbay.jetty#jetty-util;6.1.26 in local-m2-cache
        found org.mortbay.jetty#jetty-sslengine;6.1.26 in central
        found com.sun.jersey#jersey-core;1.9 in central
        found com.sun.jersey#jersey-json;1.9 in central
        found org.codehaus.jettison#jettison;1.1 in local-m2-cache
        found com.sun.xml.bind#jaxb-impl;2.2.3-1 in local-m2-cache
        found javax.xml.bind#jaxb-api;2.2.2 in local-m2-cache
        found javax.xml.stream#stax-api;1.0-2 in local-m2-cache
        found javax.activation#activation;1.1 in central
        found org.codehaus.jackson#jackson-core-asl;1.9.13 in local-m2-cache
        found org.codehaus.jackson#jackson-mapper-asl;1.9.13 in local-m2-cache
        found org.codehaus.jackson#jackson-jaxrs;1.9.13 in local-m2-cache
        found org.codehaus.jackson#jackson-xc;1.9.13 in local-m2-cache
        found com.sun.jersey#jersey-server;1.9 in central
        found asm#asm;3.2 in central
        found log4j#log4j;1.2.17 in local-m2-cache
        found net.java.dev.jets3t#jets3t;0.9.0 in central
        found org.apache.httpcomponents#httpclient;4.2.5 in local-m2-cache
        found org.apache.httpcomponents#httpcore;4.2.5 in central
        found com.jamesmurty.utils#java-xmlbuilder;0.4 in central
        found commons-lang#commons-lang;2.6 in local-m2-cache
        found commons-configuration#commons-configuration;1.6 in local-m2-cache
        found commons-digester#commons-digester;1.8 in local-m2-cache
        found commons-beanutils#commons-beanutils;1.7.0 in local-m2-cache
        found commons-beanutils#commons-beanutils-core;1.8.0 in local-m2-cache
        found org.slf4j#slf4j-api;1.7.10 in central
        found org.apache.avro#avro;1.7.4 in central
        found com.thoughtworks.paranamer#paranamer;2.3 in local-m2-cache
        found org.xerial.snappy#snappy-java;1.0.4.1 in central
        found org.apache.commons#commons-compress;1.4.1 in central
        found org.tukaani#xz;1.0 in central
        found com.google.code.gson#gson;2.2.4 in local-m2-cache
        found org.apache.hadoop#hadoop-auth;2.7.7 in central
        found org.apache.directory.server#apacheds-kerberos-codec;2.0.0-M15 in local-m2-cache
        found org.apache.directory.server#apacheds-i18n;2.0.0-M15 in local-m2-cache
        found org.apache.directory.api#api-asn1-api;1.0.0-M20 in local-m2-cache
        found org.apache.directory.api#api-util;1.0.0-M20 in local-m2-cache
        found org.apache.zookeeper#zookeeper;3.4.6 in local-m2-cache
        found org.slf4j#slf4j-log4j12;1.7.10 in central
        found io.netty#netty;3.6.2.Final in central
        found org.apache.curator#curator-framework;2.7.1 in central
        found org.apache.curator#curator-client;2.7.1 in central
        found com.jcraft#jsch;0.1.54 in spark-list
        found org.apache.curator#curator-recipes;2.7.1 in central
        found com.google.code.findbugs#jsr305;3.0.0 in local-m2-cache
        found org.apache.htrace#htrace-core;3.1.0-incubating in central
        found org.mortbay.jetty#servlet-api;2.5-20081211 in central
        found javax.servlet.jsp#jsp-api;2.1 in central
        found jline#jline;0.9.94 in central
        found junit#junit;4.11 in central
        found org.hamcrest#hamcrest-core;1.3 in central
        found com.fasterxml.jackson.core#jackson-databind;2.2.3 in central
        found com.fasterxml.jackson.core#jackson-annotations;2.2.3 in central
        found com.fasterxml.jackson.core#jackson-core;2.2.3 in central
        found com.amazonaws#aws-java-sdk;1.7.4 in central
        found joda-time#joda-time;2.10.10 in central
        [2.10.10] joda-time#joda-time;[2.2,)
        found org.scalaj#scalaj-http_2.11;2.4.2 in central
        found org.json4s#json4s-native_2.11;3.7.0-M8 in central
        found org.json4s#json4s-core_2.11;3.7.0-M8 in central
        found org.json4s#json4s-ast_2.11;3.7.0-M8 in central
        found org.json4s#json4s-scalap_2.11;3.7.0-M8 in central
        found com.thoughtworks.paranamer#paranamer;2.8 in local-m2-cache
        found com.google.guava#guava;16.0.1 in local-m2-cache
        found com.google.guava#failureaccess;1.0.1 in central
downloading /home/ariels/.ivy2/local/io.treeverse/core-247_2.11/0.1.0-SNAPSHOT/jars/core-247_2.11.jar ...
        [SUCCESSFUL ] io.treeverse#core-247_2.11;0.1.0-SNAPSHOT!core-247_2.11.jar (1ms)
downloading https://repo1.maven.org/maven2/com/thesamet/scalapb/scalapb-runtime_2.11/0.9.7/scalapb-runtime_2.11-0.9.7.jar ...
        [SUCCESSFUL ] com.thesamet.scalapb#scalapb-runtime_2.11;0.9.7!scalapb-runtime_2.11.jar (783ms)
downloading https://repo1.maven.org/maven2/com/thesamet/scalapb/lenses_2.11/0.9.7/lenses_2.11-0.9.7.jar ...
        [SUCCESSFUL ] com.thesamet.scalapb#lenses_2.11;0.9.7!lenses_2.11.jar (81ms)
downloading https://repo1.maven.org/maven2/com/google/protobuf/protobuf-java/3.8.0/protobuf-java-3.8.0.jar ...
        [SUCCESSFUL ] com.google.protobuf#protobuf-java;3.8.0!protobuf-java.jar(bundle) (263ms)
:: resolution report :: resolve 4451ms :: artifacts dl 1148ms
        :: modules in use:
        asm#asm;3.2 from central in [default]
        com.amazonaws#aws-java-sdk;1.7.4 from central in [default]
        com.fasterxml.jackson.core#jackson-annotations;2.2.3 from central in [default]
        com.fasterxml.jackson.core#jackson-core;2.2.3 from central in [default]
        com.fasterxml.jackson.core#jackson-databind;2.2.3 from central in [default]
        com.google.code.findbugs#jsr305;3.0.0 from local-m2-cache in [default]
        com.google.code.gson#gson;2.2.4 from local-m2-cache in [default]
        com.google.guava#failureaccess;1.0.1 from central in [default]
        com.google.guava#guava;16.0.1 from local-m2-cache in [default]
        com.google.protobuf#protobuf-java;3.8.0 from central in [default]
        com.jamesmurty.utils#java-xmlbuilder;0.4 from central in [default]
        com.jcraft#jsch;0.1.54 from spark-list in [default]
        com.lihaoyi#fastparse_2.11;2.1.2 from central in [default]
        com.lihaoyi#sourcecode_2.11;0.1.6 from central in [default]
        com.sun.jersey#jersey-core;1.9 from central in [default]
        com.sun.jersey#jersey-json;1.9 from central in [default]
        com.sun.jersey#jersey-server;1.9 from central in [default]
        com.sun.xml.bind#jaxb-impl;2.2.3-1 from local-m2-cache in [default]
        com.thesamet.scalapb#lenses_2.11;0.9.7 from central in [default]
        com.thesamet.scalapb#scalapb-runtime_2.11;0.9.7 from central in [default]
        com.thoughtworks.paranamer#paranamer;2.8 from local-m2-cache in [default]
        commons-beanutils#commons-beanutils;1.7.0 from local-m2-cache in [default]
        commons-beanutils#commons-beanutils-core;1.8.0 from local-m2-cache in [default]
        commons-cli#commons-cli;1.2 from local-m2-cache in [default]
        commons-codec#commons-codec;1.15 from local-m2-cache in [default]
        commons-collections#commons-collections;3.2.2 from local-m2-cache in [default]
        commons-configuration#commons-configuration;1.6 from local-m2-cache in [default]
        commons-digester#commons-digester;1.8 from local-m2-cache in [default]
        commons-httpclient#commons-httpclient;3.1 from local-m2-cache in [default]
        commons-io#commons-io;2.4 from local-m2-cache in [default]
        commons-lang#commons-lang;2.6 from local-m2-cache in [default]
        commons-logging#commons-logging;1.1.3 from local-m2-cache in [default]
        commons-net#commons-net;3.1 from local-m2-cache in [default]
        io.netty#netty;3.6.2.Final from central in [default]
        io.treeverse#core-247_2.11;0.1.0-SNAPSHOT from local-ivy-cache in [default]
        javax.activation#activation;1.1 from central in [default]
        javax.servlet#servlet-api;2.5 from central in [default]
        javax.servlet.jsp#jsp-api;2.1 from central in [default]
        javax.xml.bind#jaxb-api;2.2.2 from local-m2-cache in [default]
        javax.xml.stream#stax-api;1.0-2 from local-m2-cache in [default]
        jline#jline;0.9.94 from central in [default]
        joda-time#joda-time;2.10.10 from central in [default]
        junit#junit;4.11 from central in [default]
        log4j#log4j;1.2.17 from local-m2-cache in [default]
        net.java.dev.jets3t#jets3t;0.9.0 from central in [default]
        org.apache.avro#avro;1.7.4 from central in [default]
        org.apache.commons#commons-compress;1.4.1 from central in [default]
        org.apache.commons#commons-math3;3.1.1 from central in [default]
        org.apache.curator#curator-client;2.7.1 from central in [default]
        org.apache.curator#curator-framework;2.7.1 from central in [default]
        org.apache.curator#curator-recipes;2.7.1 from central in [default]
        org.apache.directory.api#api-asn1-api;1.0.0-M20 from local-m2-cache in [default]
        org.apache.directory.api#api-util;1.0.0-M20 from local-m2-cache in [default]
        org.apache.directory.server#apacheds-i18n;2.0.0-M15 from local-m2-cache in [default]
        org.apache.directory.server#apacheds-kerberos-codec;2.0.0-M15 from local-m2-cache in [default]
        org.apache.hadoop#hadoop-annotations;2.7.7 from central in [default]
        org.apache.hadoop#hadoop-auth;2.7.7 from central in [default]
        org.apache.hadoop#hadoop-aws;2.7.7 from central in [default]
        org.apache.hadoop#hadoop-common;2.7.7 from central in [default]
        org.apache.htrace#htrace-core;3.1.0-incubating from central in [default]
        org.apache.httpcomponents#httpclient;4.2.5 from local-m2-cache in [default]
        org.apache.httpcomponents#httpcore;4.2.5 from central in [default]
        org.apache.zookeeper#zookeeper;3.4.6 from local-m2-cache in [default]
        org.codehaus.jackson#jackson-core-asl;1.9.13 from local-m2-cache in [default]
        org.codehaus.jackson#jackson-jaxrs;1.9.13 from local-m2-cache in [default]
        org.codehaus.jackson#jackson-mapper-asl;1.9.13 from local-m2-cache in [default]
        org.codehaus.jackson#jackson-xc;1.9.13 from local-m2-cache in [default]
        org.codehaus.jettison#jettison;1.1 from local-m2-cache in [default]
        org.hamcrest#hamcrest-core;1.3 from central in [default]
        org.json4s#json4s-ast_2.11;3.7.0-M8 from central in [default]
        org.json4s#json4s-core_2.11;3.7.0-M8 from central in [default]
        org.json4s#json4s-native_2.11;3.7.0-M8 from central in [default]
        org.json4s#json4s-scalap_2.11;3.7.0-M8 from central in [default]
        org.mortbay.jetty#jetty;6.1.26 from central in [default]
        org.mortbay.jetty#jetty-sslengine;6.1.26 from central in [default]
        org.mortbay.jetty#jetty-util;6.1.26 from local-m2-cache in [default]
        org.mortbay.jetty#servlet-api;2.5-20081211 from central in [default]
        org.rocksdb#rocksdbjni;6.6.4 from central in [default]
        org.scalaj#scalaj-http_2.11;2.4.2 from central in [default]
        org.slf4j#slf4j-api;1.7.10 from central in [default]
        org.slf4j#slf4j-log4j12;1.7.10 from central in [default]
        org.tukaani#xz;1.0 from central in [default]
        org.xerial.snappy#snappy-java;1.0.4.1 from central in [default]
        xmlenc#xmlenc;0.52 from local-m2-cache in [default]
        :: evicted modules:
        com.google.guava#guava;11.0.2 by [com.google.guava#guava;16.0.1] in [default]
        commons-codec#commons-codec;1.4 by [commons-codec#commons-codec;1.15] in [default]
        com.thoughtworks.paranamer#paranamer;2.3 by [com.thoughtworks.paranamer#paranamer;2.8] in [default]
        com.google.protobuf#protobuf-java;2.5.0 by [com.google.protobuf#protobuf-java;3.8.0] in [default]
        ---------------------------------------------------------------------
        |                  |            modules            ||   artifacts   |
        |       conf       | number| search|dwnlded|evicted|| number|dwnlded|
        ---------------------------------------------------------------------
        |      default     |   88  |   5   |   4   |   4   ||   84  |   4   |
        ---------------------------------------------------------------------
:: retrieving :: org.apache.spark#spark-submit-parent-9701fd43-edf3-4326-882f-e66c357e86f5
        confs: [default]
        4 artifacts copied, 80 already retrieved (6329kB/21ms)
21/03/10 16:09:37 WARN Utils: Your hostname, redqueen resolves to a loopback address: 127.0.1.1; using 192.168.1.87 instead (on interface wlp0s20f3)
21/03/10 16:09:37 WARN Utils: Set SPARK_LOCAL_IP if you need to bind to another address
21/03/10 16:09:38 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Using Spark's default log4j profile: org/apache/spark/log4j-defaults.properties
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
21/03/10 16:09:42 WARN Utils: Service 'SparkUI' could not bind on port 4040. Attempting port 4041.
21/03/10 16:09:42 WARN Utils: Service 'SparkUI' could not bind on port 4041. Attempting port 4042.
Spark context Web UI available at http://192.168.1.87:4042
Spark context available as 'sc' (master = local, app id = local-1615385382738).
Spark session available as 'spark'.
Welcome to
      ____              __
     / __/__  ___ _____/ /__
    _\ \/ _ \/ _ `/ __/  '_/
   /___/ .__/\_,_/_/ /_/\_\   version 2.4.7
      /_/
         
Using Scala version 2.11.12 (OpenJDK 64-Bit Server VM, Java 11.0.10)
Type in expressions to have them evaluated.
Type :help for more information.

scala> import io.treeverse._
import io.treeverse._

scala> import io.treeverse.clients._; val df = LakeFSContext.newDF(spark, "...", "d1707109d96e25202d78ed9a97b..."); df.take(10)
import io.treeverse.clients._                                                   
df: org.apache.spark.sql.DataFrame = [key: string, address: string ... 3 more fields]
res0: Array[org.apache.spark.sql.Row] = Array([_$folder$,s3://...]
```

and similarly with 3.0.1, commandline
```
ariels@redqueen:~/dev/spark-client$ spark-shell --master local --packages io.treeverse:core-301_2.12:0.1.0-SNAPSHOT  --conf spark.hadoop.lakefs.api.url=http://yoni-test.lakefs.dev/api/v1  --conf spark.hadoop.lakefs.api.access_key=AKIA... --conf spark.hadoop.lakefs.api.secret_key=... --conf spark.hadoop.fs.s3a.access.key=AKIA... --conf spark.hadoop-fs.s3a.secret.key=...
```

arielshaqed committed:
Compile for 2 versions of Spark using 2 versions of Scala

- Use absolute proto pathnames so protoc can find includes in its root.
- Spark 2.4.7 provides a Scala 2.11 runtime, at least when downloaded from apache.org.  Do a
  full cross-compile.
- Downgrade some Spark 2.12 features:
  * trailing commas on last element of list,
  * implicit conversion to `java.util.concurrent.Callable`

arielshaqed committed:
Manually add timestamp.proto, and clean up sbt

arielshaqed commented:
Relevant to #21, treeverse/lakefs#1568.

arielshaqed commented:
This PR _also_ disables `examples`.  Will re-enable it in the next PR.

johnnyaug commented on build.sbt:
```diff
@@ -1,45 +1,76 @@
+// Spark versions 2.4.7 and 3.0.1 use different Scala versions.  Changing this is a deep
+// change, so key the Spark distinction by the Scala distinction.  sbt doesn't appear to
+// support other ways of changing emitted Scala binary versions using the same compiler,
+// sigh...
```
```suggestion
// support other ways of changing emitted Scala binary versions using the same compiler.
```

arielshaqed replied:
Not fun.
Done.

johnnyaug commented on build.sbt:
```diff
@@ -1,45 +1,76 @@
+    .settings(
+      sharedSettings,
+      scalaVersion := buildType.scalaVersion,
+      Compile / resourceDirectory := baseDirectory.value / ".." / ".." / "core" / "src" / "main" / "resources",
```
can this go with the common settings?

arielshaqed replied:
No, because `examples` gets different settings.  We _could_ try to generate all paths using a function, and then call a that function with different args in `core` and in `examples`.  It will avoid repeating configuration, but may make the SBT more like a program and therefore less explicit.  Pushing as a separate commit, will do whichever you prefer.

johnnyaug commented on build.sbt:
```diff
@@ -1,45 +1,76 @@
+   * therefore `./src` lives two directories above base. Also do this for `resources`
+   * etc, if needed.
+   */
+  scalaSource in Compile := baseDirectory.value / ".." / ".." / "core" / "src" / "main" / "scala",
```
Can't use the "ThisBuild" trick here instead of ".."?

johnnyaug commented on build.sbt:
```diff
@@ -1,45 +1,76 @@
 )
 
+def generateProject(buildType: BuildType) = {
+  lazy val core = Project(s"core-${buildType.name}", file(s"target/core:${buildType.name}"))
```
When I run `sbt assembly`, I'm getting an error because of the `:` in the path:
```
[error] java.lang.RuntimeException: Path '/home/yoni/dev/spark-client/target/core:301/target/scala-2.12/classes' contains separator ':'
```

(I get this error for both Scala versions).

johnnyaug commented on build.sbt:
```diff
@@ -1,45 +1,76 @@
+   * etc, if needed.
+   */
+  scalaSource in Compile := baseDirectory.value / ".." / ".." / "core" / "src" / "main" / "scala",
+  scalaSource in Test := baseDirectory.value / ".." / ".." / "core" / "src" / "test" / "scala",
```
This works for me:
```suggestion
  Compile / sourceDirectory := (ThisBuild / baseDirectory).value / "core" / "src",
  Compile / scalaSource := (ThisBuild / sourceDirectory).value / "main" / "scala",
  Test / scalaSource := (Compile / sourceDirectory).value / "test" / "scala",
```

johnnyaug commented on build.sbt:
```diff
@@ -1,45 +1,76 @@
+        scalapb.gen() -> (Compile / sourceManaged).value / "scalapb"
+      ),
+      Compile / PB.includePaths += (ThisBuild / baseDirectory).value / "core" / "src" / "main" / "resources",
+      Compile / PB.protoSources += (ThisBuild / baseDirectory).value / "core" / "src" / "main" / "resources",
```
Can also move to common settings, not sure if that's desirable

arielshaqed replied:
As above, see 7ba9d47a540ba48782b7015001b612a1a48d5f10 and let me know if you think it's better.

johnnyaug reviewed (changes requested):
Impressive work!

arielshaqed committed:
Add (separate!) examples JAR

Tested (on our staging Sonatype repo, so `--repositories` should *not* be necessary once we
publish):

```sh
spark-submit --deploy-mode cluster \
	     --repositories https://s01.oss.sonatype.org/content/repositories/staging/ \
	     --packages io.treeverse:examples-301_2.12:0.1.0-SNAPSHOT.4,io.treeverse:core-301_2.12:0.1.0-SNAPSHOT.4 \
	     --class io.treeverse.clients.examples.List \
	     --conf spark.hadoop.lakefs.api.url=http://FILL-ME.lakefs.dev/api/v1 \
	     --conf spark.hadoop.lakefs.api.access_key=AKIALAKEFS \
	     --conf spark.hadoop.lakefs.api.secret_key=lakefssecretjknsadgfjkfsdg \
	     --conf spark.hadoop.fs.s3a.access.key=AKIAS3S3S3 \
	     --conf spark.hadoop-fs.s3a.secret.key=s3secrethjbksdfagjkjkgfsd \
	     /dev/null REPO REF s3://path/to/output
```

Outputs (currently) in `s3://treeverse-ariels-test-us/du-test.2` for your edification.

arielshaqed committed:
[CR] Refactor build directory traversal cruft to single function

Less verbose, fewer copies of code, but possibly less explicit about what is being built.

arielshaqed committed:
[CR] Make comment less whiny.

arielshaqed reviewed (commented):
Thanks!

PTAL...

arielshaqed commented:
Note that it now also publishes the examples (and they work, at least on EMR).

arielshaqed committed:
[CR] Directly access directories from root

Avoid navigating from `target/{core,examples}` to the _real_ `src/` directories, instead
navigate from the root (values in `ThisBuild`).  Also re-use the value of "resources" in
all three locations.

arielshaqed commented:
PTAAL...

arielshaqed committed:
[CR] Remove dead comment

arielshaqed commented:
Thanks!
